backends/csvq.go

```diff
@@ -0,0 +1,230 @@
+package backends
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+	"log"
+	"strings"
+	"sync"
+	"time"
+)
+
+// csvqDB represents the csvqDB backend.
+type csvqDB struct {
+	SqlDB
+}
+
+// csvqDBWriter represents a writer that saves results
+// to a csvqDB backend.
+type csvqDBWriter struct {
+	sqlDBWriter
+}
+
+// NewSQLBackend returns a new csvqDB result backend instance.
+// It accepts an *sql.DB connection
+func NewCsvqBackend(db *sql.DB, opt Opt, l *log.Logger) (ResultBackend, error) {
+
+	s := csvqDB{
+		SqlDB{
+			db:              db,
+			opt:             opt,
+			resTableSchemas: make(map[string]insertSchema),
+			schemaMutex:     sync.RWMutex{},
+			logger:          l,
+		},
+	}
+
+	// Config.
+	if opt.ResultsTable != "" {
+		s.opt.ResultsTable = opt.ResultsTable
+	} else {
+		s.opt.ResultsTable = "results_%s"
+	}
+
+	return &s, nil
+}
+
+// NewResultSet returns a new instance of an csvqDB result writer.
+// A new instance should be acquired for every individual job result
+// to be written to the backend and then thrown away.
+func (s *csvqDB) NewResultSet(jobID, taskName string, ttl time.Duration) (ResultSet, error) {
+	tx, err := s.db.Begin()
+	if err != nil {
+		return nil, err
+	}
+
+	return &csvqDBWriter{
+		sqlDBWriter{
+			Writer{
+				jobID:    jobID,
+				taskName: taskName,
+				tbl:      fmt.Sprintf(s.opt.ResultsTable, jobID),
+				tx:       tx,
+			},
+			s.SqlDB,
+		},
+	}, nil
+}
+
+// RegisterColTypes registers the column types of a particular taskName's result set.
+// Internally, it translates sql types into the simpler sqlDB (SQLite 3) types,
+// creates a CREATE TABLE() schema for the results table with the structure of the
+// particular taskName, and caches it be used for every subsequent result db creation
+// and population. This should only be called once for each kind of taskName.
+func (w *csvqDBWriter) RegisterColTypes(cols []string, colTypes []*sql.ColumnType) error {
+	if w.IsColTypesRegistered() {
+		return errors.New("column types are already registered")
+	}
+
+	w.cols = make([]string, len(cols))
+	copy(w.cols, cols)
+
+	// Create the insert statement.
+	// INSERT INTO xxx (col1, col2...) VALUES.
+	var (
+		colNameHolder = make([]string, len(cols))
+		colValHolder  = make([]string, len(cols))
+	)
+	for i := range w.cols {
+		colNameHolder[i] = fmt.Sprintf(`"%s"`, w.cols[i])
+
+		// This will be filled by the driver.
+		if w.opt.DBType == DbTypePostgres {
+			// Postgres placeholders are $1, $2 ...
+			colValHolder[i] = fmt.Sprintf("$%d", i+1)
+		} else {
+			colValHolder[i] = "?"
+		}
+	}
+
+	ins := fmt.Sprintf(`INSERT INTO "%%s" (%s) `, strings.Join(colNameHolder, ","))
+	ins += fmt.Sprintf("VALUES (%s)", strings.Join(colValHolder, ","))
```

Comment:
The insert command change in csvq driver... and better than ifs... (as comented in other place) use a new backend...